src/main/java/frc/robot/subsystems/Drivetrain.java

```diff
@@ -1,65 +1,525 @@
+import edu.wpi.first.math.kinematics.SwerveDriveOdometry;
+import edu.wpi.first.math.kinematics.SwerveModulePosition;
+import edu.wpi.first.math.kinematics.SwerveModuleState;
+import edu.wpi.first.networktables.GenericEntry;
+import edu.wpi.first.wpilibj.livewindow.LiveWindow;
+import edu.wpi.first.wpilibj.shuffleboard.ShuffleboardTab;
+import edu.wpi.first.math.util.Units;
+import edu.wpi.first.wpilibj.smartdashboard.Field2d;
+import edu.wpi.first.wpilibj.smartdashboard.SendableChooser;
+import edu.wpi.first.wpilibj.smartdashboard.SmartDashboard;
 import edu.wpi.first.wpilibj2.command.SubsystemBase;
+import frc.robot.Robot;
 import frc.robot.constants.Constants;
 import frc.robot.constants.DriveConstants;
-import frc.robot.util.MotorFactory;
+import frc.robot.constants.ModuleConstants;
+import frc.robot.constants.DriveConstants.TestDriveConstants;
+import frc.robot.util.LogManager;
 
+/** Represents a swerve drive style drivetrain.
+ * Module IDs are:
+ * 1: Front left
+ * 2: Front right
+ * 3: Back left
+ * 4: Back right
+*/
 public class Drivetrain extends SubsystemBase {
 
-  private final WPI_TalonFX m_leftMotor1;
-  private final WPI_TalonFX m_leftMotor2;
-  private final WPI_TalonFX m_rightMotor1;
-  private final WPI_TalonFX m_rightMotor2;
+  // Swerve modules and other
+  public SwerveModuleState[] m_swerveModuleStates = new SwerveModuleState[] {
+    new SwerveModuleState(),
+    new SwerveModuleState(),
+    new SwerveModuleState(),
+    new SwerveModuleState()
+  };
+
+  private final Module[] m_modules;
+
+  // TODO: Should this be public?
+  public final SwerveDriveKinematics m_kinematics = new SwerveDriveKinematics(
+    new Translation2d(DriveConstants.kTrackWidth / 2, DriveConstants.kTrackWidth / 2),
+    new Translation2d(DriveConstants.kTrackWidth / 2, -DriveConstants.kTrackWidth / 2),
+    new Translation2d(-DriveConstants.kTrackWidth / 2, DriveConstants.kTrackWidth / 2),
+    new Translation2d(-DriveConstants.kTrackWidth / 2, -DriveConstants.kTrackWidth / 2)
+  );
+
+  // Pigeon
+  private final WPI_Pigeon2 m_pigeon = new WPI_Pigeon2(DriveConstants.kPigeon, Constants.kCanivoreCAN);
+  private boolean m_hasResetYaw = false; // the initial yaw has been set 
+
+  public double m_headingPIDOutput = 0;
 
-  public Drivetrain() {
+  // Odometry
+  private final SwerveDriveOdometry m_odometry;
+  private Pose2d m_robotPose = new Pose2d();
 
-    m_leftMotor1 = MotorFactory.createTalonFX(DriveConstants.kLeftMotor1, Constants.kRioCAN);
-    m_leftMotor2 = MotorFactory.createTalonFX(DriveConstants.kLeftMotor2, Constants.kRioCAN);
-    m_rightMotor1 = MotorFactory.createTalonFX(DriveConstants.kRightMotor1, Constants.kRioCAN);
-    m_rightMotor2 = MotorFactory.createTalonFX(DriveConstants.kRightMotor2, Constants.kRioCAN);
+  // Displays the field with the robots estimated pose on it
+  private final Field2d m_fieldDisplay = new Field2d();
+  
+  // PID Controllers
+  private PIDController m_xController = new PIDController(0.1, 0, 0);
```

Comment:
Should 0.1 be a constant?

Reply:
it is a dummy P value it is too small but lets you run the test

Reply:
Can you make it a constant and put  comment for it saying exactly what you said above?